docs/notebooks/burger.ipynb

```diff
@@ -0,0 +1,788 @@
+    "            The storage type, which can be StorageType.RAM, StorageType.DISK,\n",
+    "            StorageType.WORK, or StorageType.NONE.\n",
+    "        write_adj_deps : bool, optional\n",
+    "            Whether the adjoint dependency data will be stored.\n",
+    "        write_ics : bool, optional\n",
+    "            Whether the forward restart data will be stored.\n",
+    "        \"\"\"\n",
+    "        # Get the initial condition\n",
+    "        u = self.forward_work_memory[StorageType.WORK].pop(n0)\n",
+    "        if write_ics:\n",
+    "            self._store_data(u, n0, storage, write_adj_deps, write_ics)\n",
+    "        for step in range(n0, min(n1, self.model[\"max_n\"])):\n",
+    "            if write_adj_deps:\n",
+    "                self._store_data(u, step, storage, write_adj_deps, write_ics)\n",
+    "            def _residual(u_new):\n",
+    "                # C is a vector containing the nonlinear term.\n",
+    "                C = np.array(self._C(*u_new))\n",
+    "                residual = (self._M + self.model[\"dt\"] * self._K).dot(u_new) + self.model[\"dt\"] * C - self._M.dot(u)\n",
+    "                residual[0] = residual[self.model[\"nx\"] - 1] = 0\n",
+    "                return residual\n",
+    "            u_new = self._solve_newton(u, _residual)\n",
+    "            u = copy.deepcopy(u_new)\n",
+    "        step += 1\n",
+    "        if step == self.model[\"max_n\"]:\n",
+    "            self.forward_final_solution = copy.deepcopy(u_new)\n",
+    "            self.adjoint_dependency[StorageType.WORK][step] = copy.deepcopy(u_new)\n",
+    "        if (not write_adj_deps \n",
+    "           or (self.mode == \"forward\" and step < (self.model[\"max_n\"]))\n",
+    "        ):\n",
+    "            self.forward_work_memory[StorageType.WORK][step] = copy.deepcopy(u_new)\n",
+    "\n",
+    "    def adjoint(self, n0, n1, clear_adj_deps):\n",
+    "        \"\"\"Advance the adjoint solver.\n",
+    "\n",
+    "        Parameters\n",
+    "        ---------\n",
+    "        n0 : int\n",
+    "            Initial time step.\n",
+    "        n1 : int\n",
+    "            Final time step.\n",
+    "        clear_adj_deps : bool\n",
+    "            If `True`, the adjoint dependency data will be cleared.\n",
+    "        \"\"\"\n",
+    "        self.mode = \"adjoint\"\n",
+    "        if n1 <= self.model[\"max_n\"]:\n",
+    "            if n1 == self.model[\"max_n\"]:\n",
+    "                self._initialize_adjoint()\n",
+    "            u_adj = self.adjoint_work_memory[StorageType.WORK].pop(n1)\n",
+    "        M = copy.deepcopy(self._M)\n",
+    "        M[0, 1] = M[-1, -2] = 0\n",
+    "        M[0, 0] = M[-1, -1] = 1\n",
+    "        for step in range(n1, n0, - 1):\n",
+    "            J_T = self._jacobian(self.adjoint_dependency[StorageType.WORK][step]).T\n",
+    "            u_adj_coef = spsolve(J_T, u_adj)\n",
+    "            u_adj_new = M.T.dot(u_adj_coef)\n",
```

Comment:
```suggestion
    "            u_adj_coef = spsolve(J_T, u_adj)\n",
    "            u_adj_coef[0] = u_adj_coeff[-1] = 0\n",
    "            u_adj_new = M.T.dot(u_adj_coef)\n",
```
Adjoint boundary conditions, assuming we don't have a boundary control.

Reply:
Done.